dashboard/app/models/user.rb

```diff
@@ -620,8 +620,16 @@ def user_level_locked?(script_level, level)
   def next_unpassed_progression_level(script)
     user_levels_by_level = user_levels_by_level(script)
 
-    # Find the user level that we've most recently had progress on
-    user_level = user_levels_by_level.values.flatten.sort_by!(&:updated_at).last
+    # some of our user_levels may be for levels within level_groups, or for levels
+    # that are no longer in this script. we want to ignore those, and only look
+    # user_levels that have matching script_levels
+    levels = script.script_levels.map(&:level_ids).flatten
+    user_levels_by_level = user_levels.
+      where(script_id: script.id, level: levels).
+      index_by(&:level_id)
+
+    # Find the user_level that we've most recently had progress on
+    user_level = user_levels_by_level.values.max_by(&:updated_at)
```

Comment:
New approach. Instead of calling script_levels on each level (which was a db hit each time), we start out by only getting user_levels for which we have a script_level. The call to `script.script_levels` should be cached, as is the call to level_ids on each script_level.

Now, on line 636 we're guaranteed to have last_script_level by non-nil, and thus our return will be non-nil